// vistorian-core/build/src/ui/legend.js
import * as d3 from "d3";
import $ from "jquery";
export class Legend {
    constructor(data, handlerFunction) {
        this.margin = {
            left: 10,
            top: 20,
        };
        this.height = 0; // INIT ????
        this.clickCallBack = () => null; // BEFORE Function;
        this.data = data;
    }
    setClickCallBack(handlerFunction) {
        this.clickCallBack = handlerFunction;
    }
    appendTo(svg) {
        this.legendEntries = d3
            .select("#legendSvg")
            .selectAll(".legend")
            .data(this.data)
            .enter()
            .append("g")
            .attr("transform", (d, i) => {
            return ("translate(" +
                this.margin.left +
                "," +
                (this.margin.top + i * 20) +
                ")");
        });
        this.height = this.margin.top + this.data.length * 20;
        $(svg).height(this.height);
        this.legendEntries
            .append("circle")
            .attr("r", 5)
            .style("opacity", 0.7)
            .style("fill", function (d) {
            if (d.color)
                return d.color;
            else
                return "#000";
        })
            .style("stroke", function (d) {
            if (d.color)
                return d.color;
            else
                return "#000";
        })
            .style("stroke-width", 2)
            .on("click", this.clickCallBack);
        this.legendEntries
            .append("text")
            .text(function (d) {
            return d.name;
        })
            .attr("x", 20)
            .attr("y", 5);
    }
}
